### PR TITLE
Feature/attach comment to card

### PR DESCRIPTION
**Pull request by RMartin0717**
# Pull Request Template
## Description

Debug and rewrite some logic for class methods to maintain ability to implement unique ids for idea cards so that comments can be tied to their respective cards by id.

Fixes # (issue)
## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -38,22 +36,25 @@ function saveIdea(event) {
...
+  var storedIdeas = JSON.parse(localStorage.getItem("ideas"));
+  if (storedIdeas) {
+    for (var i = 0; i < storedIdeas.length; i++) {
+      //returns array as an array of instances of the idea class
+      newIdea = new Idea(storedIdeas[i].title, storedIdeas[i].body, storedIdeas[i].star, storedIdeas[i].id);
+      ideas.push(newIdea);
+    }
   }
-  
+  localStorage.setItem("ideas", JSON.stringify(ideas));
```
Comment (sme93) [inline src/main.js:58]: This line isn't needed - we only call renderPage once and it's pulling ideas from storage, it doesn't remove them from storage so we don't need to reset them.

File: src/main.js [hunk 2]
```diff
@@ -38,22 +36,25 @@ function saveIdea(event) {
...
 
 function renderPage() {
-  var retrievedIdeas = JSON.parse(localStorage.getItem("ideas"));
-  
-  for (var i = 0; i < retrievedIdeas.length; i++) {
-    newIdea = new Idea(retrievedIdeas[i].title, retrievedIdeas[i].body, retrievedIdeas[i].star, retrievedIdeas[i].id);
-    ideas.push(newIdea);
+  //handles error on load for null array
+  var storedIdeas = JSON.parse(localStorage.getItem("ideas"));
+  if (storedIdeas) {
```
Comment (sme93) [inline src/main.js:51]: Nice catch!

File: src/main.js [hunk 3]
```diff
@@ -38,22 +36,25 @@ function saveIdea(event) {
   }
 
   newIdea = new Idea(title.value, body.value, false, Date.now());
-  
-  newIdea.saveToStorage(newIdea);
+
+  newIdea.saveToStorage();
```
Comment (sme93) [inline src/main.js:41]: Not a big fan of this change, see comments in `idea.js`

File: src/idea.js [hunk 4]
```diff
@@ -6,42 +6,44 @@ class Idea  {
     this.star = star;
     this.comments = [];
   }
-  saveToStorage(idea) {
+  saveToStorage() {
```
Comment (sme93) [inline src/idea.js:12]: Both functions get retrieved ideas out of local storage, add an idea to the retrieved ideas, and push all ideas back to local storage. The old function took an idea as an argument.
New function goes through a loop, matches up ideas to find an idea and then is the same. 
The other way is way more straightforward and less confusing. I'm not sure why this change is better.

File: src/idea.js [hunk 5]
```diff
@@ -6,42 +6,44 @@ class Idea  {
...
     var retrievedIdeas = JSON.parse(localStorage.getItem("ideas"));
+    //if retrievedIdeas is empty, exit the method
     if (!retrievedIdeas) {
       return;
-    } 
-
-    for (var i = 0; i < retrievedIdeas.length; i++) {
-      if (this.id === retrievedIdeas[i].id) {
-        retrievedIdeas.splice(i, 1);
+    } else {
```
Comment (sme93) [inline src/idea.js:37]: Else here is unnecessary - where there are no ideas we return and leave the function. If there are ideas we continue. Also, we had decided to take out the if(!retrievedIdeas) statement anyway so the whole function can look like this - 
```js
deleteFromStorage() {
    var retrievedIdeas = JSON.parse(localStorage.getItem("ideas"));	    
    for (var i = 0; i < retrievedIdeas.length; i++) {
      if (this.id === retrievedIdeas[i].id) {
        retrievedIdeas.splice(i, 1);
      }
    }
    localStorage.setItem("ideas", JSON.stringify(retrievedIdeas));
  }
```

File: src/main.js (hunk 2 above)
Comment (jon-schlandt) [reply src/main.js:51]: Yep, nice error handling!

File: src/main.js [hunk 6]
```diff
@@ -112,19 +113,26 @@ function favoriteIdea(event) {
...
 
 function deleteIdea(event) {
   if (event.target.id === "deleteButton") {
     var ideaId = event.target.closest("article").id;
-    
-    ideas[ideaId].deleteFromStorage();
-    ideas.splice(ideaId, 1); 
-    
+
+    for (var i = 0; i < ideas.length; i++) {
```
Comment (sme93) [inline src/main.js:149]: This adds a double loop to our app, which isn't the worst thing, but isn't needed.

File: src/idea.js [hunk 7]
```diff
@@ -6,42 +6,44 @@ class Idea  {
...
     var retrievedIdeas = JSON.parse(localStorage.getItem("ideas"));
-    if (!retrievedIdeas) {
-      retrievedIdeas = [];
-    } 
-    retrievedIdeas.push(idea);
-    localStorage.setItem("ideas", JSON.stringify(retrievedIdeas));
+    if (retrievedIdeas.length) {
+      for (var i = 0; i < retrievedIdeas.length; i++) {
+        //if retrievedIdeas contains any ideas and the card id already exists, update the card and then update storage
+        if (retrievedIdeas[i].id === this.id) {
```
Comment (jon-schlandt) [inline src/idea.js:17]: Just wondering if this is necessary as when we are saving a new idea, I'm not sure if there will be an instance in which an idea exists with that exact id already (if the same card already exists, then there might be a problem with our delete function). But if I'm not looking with a wide enough perspective please let me know!

File: src/main.js [hunk 8]
```diff
@@ -112,19 +113,26 @@ function favoriteIdea(event) {
   if (event.target.id === "favoriteButton") {
     var ideaId = event.target.closest("article").id;
 
-    ideas[ideaId].updateIdea();
-    //ideas[ideaId].saveToStorage(ideas[ideaId]);
+  for (var i = 0; i < ideas.length; i++) {
```
Comment (sme93) [inline src/main.js:135]: Same issue as deleteId - we are complicating these functions. If there's a compelling reason we can change it.

File: src/main.js [hunk 9]
```diff
@@ -63,22 +64,22 @@ function render(arrayToRender) {
     for (var i = 0; i < arrayToRender.length; i++) {
       var imageSource = getImageSourceFromIdea(arrayToRender[i]);
         markup += `
-        <article class="idea" id="${i}">
+        <article class="idea" id="${arrayToRender[i].id}">
```
Comment (sme93) [inline src/main.js:68]: I know you need the specific id in order to attach it to a comment, I recommend we use a data attribute. I used one in a previous project that I can show you and it worked really well. 
Here's a link https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes
We can add any extra info we need using data attributes, solving the ID problem.

**Review by sme93 — CHANGES_REQUESTED** [review]
Not sure all of this change is worth it, some good stuff in here. The only reason I know about the double loop situation is because I was trying to do the same thing yesterday, and that's how I got to the solution from yesterday.

File: src/main.js (hunk 1 above)
Comment (RMartin0717) [reply src/main.js:58]: Good call--just deleted it

File: src/main.js (hunk 2 above)
Comment (RMartin0717) [reply src/main.js:51]: thanks!

File: src/idea.js (hunk 7 above)
Comment (RMartin0717) [reply src/idea.js:17]: This is so that when the idea is starred or commented, it updates (which is why it would have the same id)
